File: backend/app/schemas/platform.py
"""Platform context contracts for tenant/account scoped requests."""

from typing import Literal

from pydantic import BaseModel, Field


RoleView = Literal["trader", "investor", "admin"]
Visibility = Literal["private", "tenant_shared", "public"]
DataScope = Literal["public", "tenant", "user", "account"]
TradeMode = Literal["paper", "live"]
BrokerAdapter = Literal["paper", "xtquant_qmt", "broker_rest"]


class PlatformContext(BaseModel):
    """Resolved request scope used before querying user-owned data."""

    tenant_id: str = Field(description="Tenant boundary for this request")
    user_id: int = Field(description="Authenticated user id")
    role: str = Field(description="Raw RBAC role")
    role_view: RoleView = Field(description="Product persona derived from RBAC")
    account_id: str | None = Field(
        default=None,
        description="Trading account boundary for portfolio/order data",
    )
    visibility: Visibility = Field(
        default="private",
        description="Whether produced data is private, tenant-shared, or public",
    )
    data_scope: DataScope = Field(
        default="account",
        description="Smallest data ownership scope needed by the request",
    )
    cross_tenant: bool = Field(
        default=False,
        description="True when an admin explicitly works outside its default tenant",
    )

File: frontend/src/types/platform.ts
import type { Role, User } from '../contexts/AuthContext'

export type RoleView = 'trader' | 'investor' | 'admin'
export type Visibility = 'private' | 'tenant_shared' | 'public'
export type DataScope = 'public' | 'tenant' | 'user' | 'account'
export type TradeMode = 'paper' | 'live'
export type BrokerAdapter = 'paper' | 'mock_qmt' | 'xtquant_qmt' | 'broker_rest'

export interface PlatformScope {
  tenantId: string
  ownerUserId?: string
  accountId?: string
  visibility: Visibility
  dataScope: DataScope
}

export interface PlatformSession extends PlatformScope {
  roleView: RoleView
  tenantName: string
  userName: string
  tradeMode: TradeMode
  brokerAdapter: BrokerAdapter
  cloudReady: boolean
}

export function roleToRoleView(role: Role): RoleView {
  if (role === 'admin') return 'admin'
  if (role === 'user') return 'investor'
  return 'trader'
}

export function buildPlatformSessionFromUser(user: User | null): PlatformSession {
  const role = user?.role ?? 'user'
  const roleView = roleToRoleView(role)
  const fallbackTenantId = roleView === 'admin' ? 'platform' : 'tenant-default'

  return {
    tenantId: user?.tenantId || fallbackTenantId,
    tenantName: user?.tenantName || (roleView === 'admin' ? '平台运营' : '默认租户'),
    ownerUserId: user ? String(user.id) : undefined,
    accountId: user?.defaultTradeAccountId || (roleView === 'admin' ? undefined : 'paper-default'),
    visibility: roleView === 'admin' ? 'tenant_shared' : 'private',
    dataScope: roleView === 'admin' ? 'tenant' : 'account',
    roleView,
    userName: user?.name || '未登录用户',
    tradeMode: user?.tradeMode || 'paper',
    brokerAdapter: user?.brokerAdapter || 'paper',
    cloudReady: true,
  }
}
